hal/lib/athena/handles/LimitedHandleResource.h

```diff
@@ -39,42 +36,29 @@ class LimitedHandleResource {
  public:
   LimitedHandleResource(const LimitedHandleResource&) = delete;
   LimitedHandleResource operator=(const LimitedHandleResource&) = delete;
-  LimitedHandleResource();
-  THandle Allocate(const TStruct& toSet);
-  TStruct Get(THandle handle, int32_t* status);
+  LimitedHandleResource() = default;
+  THandle Allocate();
+  std::shared_ptr<TStruct> Get(THandle handle);
   void Free(THandle handle);
 
  private:
-  TStruct m_structures[size];
-  bool m_allocated[size];
+  std::shared_ptr<TStruct> m_structures[size];
   priority_mutex m_handleMutexes[size];
   priority_mutex m_allocateMutex;
 };
 
 template <typename THandle, typename TStruct, int16_t size,
           HalHandleEnum enumValue>
-LimitedHandleResource<THandle, TStruct, size,
-                      enumValue>::LimitedHandleResource() {
-  // ensure all allocations are initially false.
-  for (int i = 0; i < size; i++) {
-    m_allocated[i] = false;
-  }
-}
-
-template <typename THandle, typename TStruct, int16_t size,
-          HalHandleEnum enumValue>
-THandle LimitedHandleResource<THandle, TStruct, size, enumValue>::Allocate(
-    const TStruct& toSet) {
+THandle LimitedHandleResource<THandle, TStruct, size, enumValue>::Allocate() {
   // globally lock to loop through indices
   std::lock_guard<priority_mutex> sync(m_allocateMutex);
   int16_t i;
   for (i = 0; i < size; i++) {
-    if (m_allocated[i] == false) {
+    if (m_structures[i] == nullptr) {
       // if a false index is found, grab its specific mutex
       // and allocate it.
       std::lock_guard<priority_mutex> sync(m_handleMutexes[i]);
-      m_allocated[i] = true;
-      m_structures[i] = toSet;
+      m_structures[i] = std::make_shared<TStruct>();
```

Comment:
Since we hold the allocateMutex for the entire duration of both free and allocate, it is not necessary to double check as the status of that vector cannot be changed.